src/core/qgsvectorlayerutils.h

```diff
@@ -167,6 +167,56 @@ class CORE_EXPORT QgsVectorLayerUtils
      * \since QGIS 3.4
      */
     static std::unique_ptr<QgsVectorLayerFeatureSource> getFeatureSource( QPointer<QgsVectorLayer> layer ) SIP_SKIP;
+
+    /**
+     * Matches the attributes in \a feature to the specified \a fields.
+     *
+     * This causes the attributes contained within the given \a feature to be rearranged (or in
+     * some cases dropped) in order to match the fields and order indicated by \a fields.
+     *
+     * \since QGIS 3.4
+     */
+    static void matchAttributesToFields( QgsFeature &feature, const QgsFields &fields );
+
+    /**
+     * Converts input \a feature to be compatible with the given \a layer.
+     *
+     * This function returns a new list of transformed features compatible with the input
+     * layer, note that the number of features returned might be greater than one.
+     *
+     * The following operations will be performed to convert the input features:
+     *  - convert single geometries to multi part
+     *  - drop additional attributes
+     *  - drop geometry if layer is geometry-less
+     *  - add missing attribute fields
+     *  - add back M/Z values (initialized to 0)
+     *  - drop Z/M
+     *  - convert multi part geometries to single part
+     *
+     * \since QGIS 3.4
+     */
+    static const QgsFeatureList makeFeatureCompatible( const QgsFeature &feature, const QgsVectorLayer &layer );
+
+    /**
+     * Converts input \a features to be compatible with the given \a layer.
+     *
+     * This function returns a new list of transformed features compatible with the input
+     * layer, note that the number of features returned might be greater than the number
+     * of input featurers.
```

Comment:
in case multipart need to be converted to single part features?

Reply:
Yes, exactly

Reply:
typo on "features"

Reply:
The question was more "does one need to read the code to find out or should it be part of the API docs" ;)